### PR TITLE
Hide fallback login link when handling OIDC callback

### PR DESCRIPTION
**Pull request by ousmaneo**
Before this change "login with default method" was being shown while handling login on OIDC callback. Now we make sure it's not displayed and show only on first load or when we have an error 
Fix Graylog2/graylog-plugin-enterprise#2385

/jenkins-pr-deps Graylog2/graylog-plugin-enterprise#2677
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Review by edmundoa — CHANGES_REQUESTED** [review]
Thank you for working on this! Unfortunately I have two concerns with these changes:

1. Thus far all OIDC/Okta code is in a plugin and here we are breaking that by introducing some OIDC specific logic into the core project. I don't think this is the right way of fixing the issue, and we should look for a more generic way (see 2).

2. This only addresses part of the issue and the text is still visible when logging out:

https://user-images.githubusercontent.com/716185/132013718-a10d3077-fb04-4c41-bd33-570ef22d1cc4.mp4

---

As for an idea on how a more generic approach would look like, I can see two viable options right now:

- Let the login form decide when that message should be displayed. This is probably the easiest way but has the drawback of losing all control over when that link will be displayed, or if it will be displayed at all when using a plugin.

- Let the login form report back its state and use that to control when the link is displayed. This would let us keep control on when the link is displayed, and defining the right states would also let us do things like showing the fallback link when the plugin first loads (e.g. right now an admin could still log in even if the OIDC plugin does not load). The hardest part is to check which states we can discern during the different loading scenarios, and define those states so they are flexible enough but also reusable.

What do you think? Do you have any other approaches that we should consider?

**Comment by ousmaneo** [comment]
> What do you think? Do you have any other approaches that we should consider?
Hi @edmundoa, thank you for the review.
My first approach was similar to your 1st point and I was confronted with the fact that the link may not be present at all in case the plugin does not load. So no way to log in.

I thought about passing the setUseFallback to the login form and let it report back on whether the link should show or not. 
I will work on this again.

File: graylog2-web-interface/src/util/URLUtils.ts [hunk 1]
```diff
@@ -111,8 +111,14 @@ const URLUtils = {
 
     return isValid;
   },
+  isCallbackUrl() {
+    const uri = new URI(window.location.href);
+
+    return uri.path() === '/authorization-code/callback';
```
[the remark this answers is not in the document]
Comment (ousmaneo) [reply graylog2-web-interface/src/util/URLUtils.ts:117]: That's correct. We only support one URL. 
I think we should open a ticket and enforce it as default value. 👍

**Comment by edmundoa** [comment]
> I thought about passing the setUseFallback to the login form and let it report back on whether the link should show or not. I think that goes in the direction of your second suggestion.

That could work, but would still leave control on whether the link is displayed or not to the plugin. My line of thinking was to define a set of states (e.g. `INITIALIZING`, `INITIALIZED`, `TRANSITIONING`, etc.) that core defines and plugins should use to report back their state. That way core could decide which states should display the link or not. What I am currently unsure is if we can detect when the user is logging out, since we redirect to `/`.

Anyway, I'm all for trying the simplest solution and see if that works.

>I will investigate more and provide a better fix.

Cool, thank you @ousmaneo!

**Comment by ousmaneo** [comment]
> > I thought about passing the setUseFallback to the login form and let it report back on whether the link should show or not. I think that goes in the direction of your second suggestion.
> 
> That could work, but would still leave control on whether the link is displayed or not to the plugin. My line of thinking was to define a set of states (e.g. `INITIALIZING`, `INITIALIZED`, `TRANSITIONING`, etc.) that core defines and plugins should use to report back their state. That way core could decide which states should display the link or not. What I am currently unsure is if we can detect when the user is logging out, since we redirect to `/`.
> 
> Anyway, I'm all for trying the simplest solution and see if that works.
> 
> > I will investigate more and provide a better fix.
> 
> Cool, thank you @ousmaneo!

Hi Edmundo, thank you for the suggestion I think it's clearer with the states.
I refactored and updates the code to use that.

**Review by edmundoa — APPROVED** [review]
It looks good to me and I think the approach looks good, great job! 👍 

The only thing I would consider adding is some documentation to each state (just a comment in the constants file would suffice imho). I'll let the PR open in case you want to add it. Feel free to merge otherwise.

**Pull request merged**